feat: Support for using multiple databases in datastore

danieljbruce opened the pull request:
This PR will add multi db support so that datastore users can interact with multiple databases.

danieljbruce committed:
A simple test with multidb support

danieljbruce committed:
A simple test with multidb support

Adds databaseId to the request options, client options and code for carrying over the client options to the request options as well as a test to make sure it actually works.

danieljbruce committed:
Merge branch 'multi-db' of https://github.com/danieljbruce/nodejs-datastore into multi-db

danieljbruce committed:
Add additional checks with get

The test should include additional checks to ensure get is running the way it is supposed to.

danieljbruce committed:
Add a test for saving with another database

danieljbruce committed:
other datastore should delete the entity

Since we are saving things to another datastore we need to delete the data from that datastore.

danieljbruce committed:
Add a test for the getDatastore method

The user needs a way to get the database Id from the client so that they can keep track of their clients more easily.

danieljbruce committed:
Add a unit test for database id

Ensure that the database id gets passed all the way down to the generated layer and gets included in the request.

danieljbruce committed:
Do an additional check for the sake of tests

We should check for the existance of datastore.options so that the code doesn’t fail if they are not provided at all.

danieljbruce committed:
initialize generated client so it can be mocked

The test on kokoro is failing because in that environment a project id isn’t defined. The solution is to eliminate the get request so that a call never reaches the backend and initialize the gapic client so it can be mocked out and then catch any call that reaches the mock.

danieljbruce committed:
comments and setDatabaseId

Add docs comments for newly added functions and add a test that ensures setDatabaseId works correctly.

danieljbruce committed:
Refactor of database name

Don’t use the same string multiple times. When the database name we are using for testing changes then we should only have to make that change in one place.

danieljbruce committed:
correction to test against old database

This was a typo. We want to test against the database we were using before

danieljbruce committed:
getRequestWithDatabaseId function

Add the getRequestWithDatabaseId function and change it so that it can easily be used in multiple places and the tests still pass.

danieljbruce committed:
Change function name to addDatabaseIdToRequest

The request options should contain databaseId so rename the function as appropriate.

danieljbruce committed:
Comments in all the places to add reqOpts

This commit contains a comment in all the places we want to add the reqOpts change.

danieljbruce committed:
Add a warning to a certain test for easier debug

Add a warning because the way the assert checking was done before, the test suite was not receiving any indication of which test was failing which makes debugging really hard.

danieljbruce committed:
This fixed the bug for the save operation

danieljbruce committed:
datastore request id

Add request ids back in all the places they are required

danieljbruce committed:
addDatabaseIdToRequest

Add addDatabaseIdToRequest to the transaction test framework so that it can be used in the tests.

danieljbruce committed:
Solve all the test errors for createReadStream

Mock out add database id for the tests in createReadStream that correspond to createReadStream.

danieljbruce committed:
addDatabaseIdToRequest mock for runQueryStream

Add the addDatabaseIdToRequest mock to the tests for every runQueryStream test so that the tests pass.

danieljbruce committed:
By default, mock out addDatabaseIdToRequest

addDatabaseIdToRequest should be a default function in datastore object of request so that we don’t have to mock it out in many different places.

danieljbruce committed:
Revert "addDatabaseIdToRequest mock for runQueryStream"

This reverts commit a3bdb926659beaf18d500d955db7a3b9ca0d7f6a.

danieljbruce committed:
Revert "Solve all the test errors for createReadStream"

This reverts commit 7b5f90289bb12d6ac81b91777a08c35f54112771.

danieljbruce committed:
linting

lint an import

danieljbruce committed:
Revert "datastore request id"

This reverts commit f167ce3447a42b66635461cb50262bac13ba74f1.

danieljbruce committed:
Revert "This fixed the bug for the save operation"

This reverts commit dd7b06ca866ae929d4dad56878b6c85c8b2155e6.

danieljbruce committed:
Revert "Comments in all the places to add reqOpts"

This reverts commit 8ba512ad130c11cdb03bd2134d932ec6cd9e116e.

danieljbruce committed:
Introduce addDatabaseIdRequest in request

Introduce adding the databaseId in request layer again.

danieljbruce committed:
Declare a constant so that tests don’t run

Importing test into the system-test folder causes the system-test command to also run the tests. Eliminate this import.

danieljbruce committed:
Remove setDatabaseId

Remove setDatabaseId as it provides a way for the user to get confused about the current state of the client.

danieljbruce committed:
Add comment for addDatabaseIdToRequest function

The function needs a comment to be consistent with docs.

danieljbruce committed:
Address header issue

The new util file needs a header

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into multi-db

danieljbruce commented on src/request.ts:
```diff
@@ -995,6 +995,8 @@ class DatastoreRequest {
       }
     }
 
+    datastore.addDatabaseIdToRequest(reqOpts);
```
This is all we need to add the database id to every request.

danieljbruce committed:
Add databaseId for encoding and decoding keys

The databaseId should be encoded on keys so that the user can write on a record by record basis

danieljbruce committed:
Revert "Add databaseId for encoding and decoding keys"

This reverts commit 0ebabd554a634a06ee8893082deeaf256b6a0337.

danieljbruce committed:
Try a test with the namespace

Try out a new test that looks at data details

danieljbruce committed:
Use the other datastore in the test

Use another datastore with key

danieljbruce committed:
Create a test that looks at specific details

The test should look at specific details in the key and should return the data with the right author.

danieljbruce committed:
Add additional checks to existing test

Add additional checks to existing test to show default database is unaffected.

danieljbruce committed:
Revert "Revert "Add databaseId for encoding and decoding keys""

This reverts commit 6449d10733c7868bcba0b2d2947551fab5df9751.

danieljbruce committed:
Revert "Revert "Revert "Add databaseId for encoding and decoding keys"""

This reverts commit e694341c9f8179bde3831f2bd9c040d7062e0496.

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into multi-db

# Conflicts:
#	system-test/datastore.ts

danieljbruce committed:
Rename the second database to secondDatabase

danieljbruce committed:
Use the database called multidb-test

multidb-test is the name of the database set up on the project used in Github automation so we need to change this name in code to use that new database.

kolea2 commented on src/index.ts:
```diff
@@ -694,10 +695,29 @@ class Datastore extends DatastoreRequest {
+   *
+   * @param {any} [reqOpts] The request operations to add the databaseId to
+   */
+  addDatabaseIdToRequest(reqOpts: any): void {
```
this looks more like an internal implementation detail. Do we need this on the surface layer?

danieljbruce replied:
I guess if we follow the plan on the second comment then this will not be needed on the surface layer anymore.

kolea2 commented on src/util.ts:
```diff
@@ -0,0 +1,19 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export function addDatabaseIdToRequest(datastore: any, reqOpts: any) {
```
this looks like it's only used in one place in the (non test) code. Is this necessary, or can we simply inline it?

danieljbruce replied:
I guess we could inline it, move this to `test/util.ts` so that multiple test files can use this shared code and then add a comment to say that the same logic is in two places.

danieljbruce replied:
Second thought, I really think we should avoid having duplicate code in two places (this logic in the test code and this logic in request.ts). So I think we should keep this function here and call it directly from both places instead of going through `addDatabaseIdToRequest` on `index.ts`.

kolea2 replied:
Will add some more comments in tests specifically

danieljbruce committed:
Merge branch 'main' into multi-db

danieljbruce committed:
Eliminate function on index that adds the id

Currently a function is used as a passthrough to the function on the util file. This unnecessarily exposes a function to the user that they shouldn’t be concerned with.

danieljbruce committed:
Fix the test so that the assertion error bubbles

The assertion error should bubble up to the user if we send it through the callback.

danieljbruce committed:
Merge branch 'main' into multi-db

kolea2 commented on test/transaction.ts:
```diff
@@ -61,6 +62,9 @@ describe('Transaction', () => {
     request_() {},
     projectId: PROJECT_ID,
     namespace: NAMESPACE,
+    addDatabaseIdToRequest(d: any, r: any) {
```
I would expect this to be: `databaseId: DATABASE` similar to project and namespace. Can you help me understand this better?

kolea2 commented on src/request.ts:
```diff
@@ -1155,6 +1158,7 @@ export interface SharedQueryOptions {
   };
 }
 export interface RequestOptions extends SharedQueryOptions {
+  databaseId?: string;
```
why is this here and not along side wherever `reqOpts.projectId` lives? (looking at `request.ts`, line 1030)

danieljbruce replied:
`databaseId` is in `RequestOptions` because `reqOpts` is of type `RequestOptions` and `reqOpts` is what gets sent [here](https://github.com/googleapis/nodejs-datastore/blob/4caf162910b2aa2a8f59c9ae26f05d69bab8b7b2/src/request.ts#L1033) to the backend.

`projectId` lives in `sharedQueryOptions` so therefore also lives in `RequestOptions` by inheritance which is why we can attach `projectId` on line 1030. Now the reason `databaseId` should not live in `sharedQueryOptions` is because then they would be included in `RunAggregationQueryRequest`, but they do not need to be. I am not sure why `projectId` lives in `sharedQueryOptions` and perhaps that is something that needs to be investigated further.

kolea2 replied:
discussed offline, will move `databaseId` to shared query options to match proto: example https://github.com/googleapis/googleapis/blob/master/google/datastore/v1/datastore.proto#L208-L234

danieljbruce committed:
Add types so that purpose of function is clear

Types currently set to any should be more specific. This will help clarify the purpose of the function that is used.

danieljbruce committed:
Merge branch 'main' into multi-db

danieljbruce committed:
Add types and names so that the function works

The types should be more specific and the names should be more explicit and clear.

danieljbruce committed:
Merge branch 'multi-db' of https://github.com/danieljbruce/nodejs-datastore into multi-db

danieljbruce committed:
Eliminate repeated code fragments

The test case has the same lines for each secondary database key and these lines should not be repeated so we regroup them into better data structures for the job

danieljbruce committed:
Better names and code organization

The names of some variables were changed to align with others and code was moved so that code pertaining to the same function is grouped together

danieljbruce committed:
Change the variable name of the default key

There is only one default key so we change the variable name so that it doesn’t mention 1

danieljbruce committed:
Refactor the post key hierarchy out

Refactor the hierarchy out for a cleaner test

danieljbruce committed:
Add only to the test we are interested in

danieljbruce committed:
Revert "Add only to the test we are interested in"

This reverts commit 0378dead81c405a85fc7bb288889fbca149cf780.

danieljbruce committed:
Merge branch 'main' into multi-db

MathiasWP commented:
Looking forward to this feature!! 🙌

kolea2 commented on test/transaction.ts:
```diff
@@ -61,6 +63,9 @@ describe('Transaction', () => {
     request_() {},
     projectId: PROJECT_ID,
     namespace: NAMESPACE,
+    addDatabaseIdToRequest(datastore: Datastore, reqOpts: RequestOptions) {
```
it's still unclear to me why we can't instead add `databaseId: DATABASE_ID` here similar to these other values (projectId, namespace). Why is database id treated differently?

danieljbruce replied:
I think I misunderstood the question from before. We could add **databaseId: SECOND_DATABASE_ID** here. DATABASE_ID is used lower on the stack than `request` and request is mocked out so it won't do anything, but it won't hurt either.

danieljbruce replied:
I cherry picked the changes with test parameterization now so that we run tests against both the default database and the second database.

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into multi-db

# Conflicts:
#	src/index.ts
#	system-test/datastore.ts

danieljbruce committed:
Add second database id

In the transaction tests we add the databaseId parameter. The databaseId parameter will run tests against the secondary database.

danieljbruce committed:
Remove unnecessary mocks

The mocks for addDatabaseIdToRequest are not necessary for these two files since they were only useful for the way the code was written before so should be removed.

danieljbruce committed:
Remove unused imports

Imports that were used in old mocks no longer exist in the file so should be removed.

danieljbruce committed:
Add parameterized testing to the test/index file

Parameterized testing can be added to the test/index file to test to see how the tests behave with databaseId set to the second database and without this setting too.

danieljbruce committed:
Add parameterized testing to test/transaction.ts

Parameterized testing should be used on transaction.ts. This commit adds the testing with async.

danieljbruce committed:
This ensures that the tests still run on the deft

Tests need to run on the default database as well as the secondary database. This commit ensures that they do.

danieljbruce committed:
Add parameterized testing for the system tests

Parameterized testing for the system tests will ensure that tests run with the default database as well as with the secondary database.

danieljbruce committed:
Use default datastore variable in new tests

The intent of the datastore variable in the new tests is to be the default database so that we are able to see if reads/writes affect the default or secondary database in these tests. They should not use the parameterized version of datastore.

kolea2 commented on system-test/datastore.ts:
```diff
@@ -17,20 +17,33 @@ import {readFileSync} from 'fs';
+        namespace: `${Date.now()}`,
+      },
+      {
+        databaseId: SECOND_DATABASE_ID,
```
where is the default db created?

danieljbruce replied:
I'm not sure I understand the question. The default db is already set up in the integration test project. This object here with databaseId and namespace is one of the parameters for the parameterized testing. The other parameter which is above it just contains namespace so in that case, when these options are used to initialize datastore then the default database is assumed to be used.

kolea2 commented on src/index.ts:
```diff
@@ -67,6 +67,7 @@ import {Transaction} from './transaction';
 import {promisifyAll} from '@google-cloud/promisify';
 import {google} from '../protos/protos';
 import {AggregateQuery} from './aggregate';
+import {addDatabaseIdToRequest} from './util';
```
per the linter, can we remove this?

kolea2 commented on src/request.ts:
```diff
@@ -992,6 +993,8 @@ class DatastoreRequest {
       }
     }
 
+    addDatabaseIdToRequest(datastore, reqOpts);
```
I think this can now be inlined

danieljbruce replied:
Yup. This can be inlined now.

danieljbruce committed:
inline add database id

Now that the tests don’t need to mock out adding database id function, we can inline the code that adds the database id to the request options.

kolea2 commented on system-test/datastore.ts:
```diff
@@ -17,20 +17,33 @@ import {readFileSync} from 'fs';
+      },
+      {
+        databaseId: SECOND_DATABASE_ID,
+        namespace: `${Date.now()}`,
```
nit - slightly update this so that we can easily distinguish between the default namespace and the secondary db namespace; maybe something like `second-db-[date]`

danieljbruce replied:
Done

danieljbruce committed:
Modify the namespace in parameterized tests

Parameterized tests should have a different namespace for the second set of tests for sure so we add this prefix.

danieljbruce committed:
Move databaseId to SharedQueryOptions for alignme

The databaseId should be moved to shared query options as discussed.

ruyadorno commented on test/transaction.ts:
```diff
@@ -50,19 +52,24 @@ const fakePfy = Object.assign({}, pfy, {
+async.each(
+    [{}, {databaseId: SECOND_DATABASE_ID}],
+    (clientOptions: DatastoreOptions) => {
 describe('Transaction', () => {
```
oh, there's a setting in the GitHub interface to do just that. Clicking the small cog icon when in the "Files Changed" tab will open the popup where it's possible to select & click on "Hide whitespace > Apply and reload" option. Using that option will give you just the same type of view you have right now _after_ you run the linter.
<img width="1087" alt="Screenshot 2023-09-27 at 4 39 15 PM" src="https://github.com/googleapis/nodejs-datastore/assets/220900/0dbe01b2-f675-410a-8d75-dfb516f6d599">

It's also a git command line option: `git diff -w`

danieljbruce replied:
Nice!

ruyadorno reviewed (approved):
LGTM 👍

danieljbruce committed:
Run the linter

Linting was broken so that the PR could be read. Ran the linter again.

danieljbruce committed:
Merge branch 'main' into multi-db
